### PR TITLE
Expose extension API (step 2)

### PR DESCRIPTION
**Pull request by pitrou**
Based on PR #1581. Contents (in addition to various amounts of yak-shaving):

* Handle a minimal Series-like object (without indexing, since that requires hash-based lookups), including compatibiliy with Numpy ufuncs (e.g. calling `np.sin` on a Series returns a new Series with the same Index)
* Break out implementation of attribute-setting in `numba.targets` to allow for easier extension
* Replace the `builtin` / `implement` pair with `lower_builtin`
* Replace the `builtin_getattr` / `impl_getattr[_generic]` pairs with `lower_getattr` and `lower_getattr_generic`
* Replace the `builtin_setattr` / `impl_setattr[_generic]` pairs with `lower_setattr` and `lower_setattr_generic`
* Rename the typing decorators `@builtin[_XXX]` to `@infer[_XXX]`
* Allow `@infer_global` as a decorator on a template class, making it easier to type global functions
* Add `@overload_method` to easily implement methods

### CODE REVIEW

**Comment by seibert** [comment]
I'm not sure if the commit history has now gone horribly awry with the commit squashing on #1581, but if you can merge master, hopefully the diff will shrink substantially.

**Comment by pitrou** [comment]
I've now merged master into this. Note there's still many changes from massaging some internal APIs.

File: numba/extending.py [hunk 1]
```diff
@@ -41,7 +42,23 @@ def generic(self):
...
+    A decorator marking the decorated function as typing and implementing
+    *func* in nopython mode.
+
+    The decorated function will have the same formal parameters as *func*
+    and be passed the Numba types of those parameters.  It should return
+    a function implementing *func* for the given types.
+
+    Here is an example implementing len() for tuple types::
+
+        @overlay(len)
```
Comment (stuartarchibald) [inline numba/extending.py]: s/overlay/overload/

File: numba/jitclass/base.py [hunk 2]
```diff
@@ -309,6 +271,7 @@ def generic(self, args, kws):
             return types.BoundFunction(MethodTemplate, instance)
 
         elif attr in instance.jitprops:
+            # It's a jitted property => typeinfer its getter
```
Comment (stuartarchibald) [inline numba/jitclass/base.py]: Thanks for adding comments like this, it makes the code easier to follow.

File: numba/npyufunc/dufunc.py [hunk 3]
```diff
@@ -218,9 +218,8 @@ def _install_cg(self, targetctx=None):
             targetctx = self._dispatcher.targetdescr.target_context
         _any = types.Any
         _arr = types.Array
-        sig0 = _any(*((_any,) * self.ufunc.nin + (_arr,) * self.ufunc.nout))
-        sig1 = _any(*((_any,) * self.ufunc.nin))
-        targetctx.insert_func_defn([(self._lower_me, [
-            (self, sig0),
-            (self, sig1),
-        ])])
+        # Either all outputs are explicit or none of them is
```
Comment (stuartarchibald) [inline numba/npyufunc/dufunc.py]: s/is/are/ ?

File: numba/targets/imputils.py [hunk 4]
```diff
@@ -16,130 +17,137 @@ class Registry(object):
...
-        while hasattr(curr_item, '__wrapped__'):
-            self.attributes.append(curr_item)
-            curr_item = curr_item.__wrapped__
-        return item
-
-    def register_cast(self, impl, sig):
-        self.casts.append((impl, sig))
+    def lower(self, func, *argtys):
+        """
+        Decorate an implementation of *func* for the given argument types.
```
Comment (stuartarchibald) [inline numba/targets/imputils.py]: Perhaps `Use to decorate...`

File: numba/targets/imputils.py [hunk 5]
```diff
@@ -16,130 +17,137 @@ class Registry(object):
...
+        *func* may be an actual global function object, or any
+        pseudo-function supported by Numba, such as "getitem".
 
+        The decorated implementation has the signature
```
Comment (stuartarchibald) [inline numba/targets/imputils.py]: Perhaps `The decorated implementation will have the signature` ?

File: numba/targets/imputils.py [hunk 6]
```diff
@@ -8,6 +8,7 @@
 import functools
```
Comment (stuartarchibald) [inline numba/targets/imputils.py]: The two comments below relating to docstrings:
https://github.com/numba/numba/pull/1599/files#r48951208
https://github.com/numba/numba/pull/1599/files#r48951240
, if considered valid, are applicable to most function definitions that are decorators within this file.

**Comment by stuartarchibald** [comment]
I've reviewed the current state of the code (64bba13). I'm of the view that these changes contribute significantly to the clarity of the code base. The removal of `@builtin*` for a more specific set of decorator names will make this considerably easier to document and in the long term make it easier for new users to understand. The only further change could be the adjustment of the `@builtin` name in the typing module, as discussed on IRC.

File: numba/targets/imputils.py (hunk 5 above)
Comment (pitrou) [inline numba/targets/imputils.py]: Ok, I've changed to future tense.

**Comment by seibert** [comment]
Alright, thanks for the review everyone!

**Pull request merged**